docs/DesignDocs/ExtendedEnclaveInitializationData.md

```diff
@@ -0,0 +1,168 @@
+    }
+```
+
+Then, after the enclave image is loaded but before it is initialized, we add the additional memory pages containing the contents of the `oe_eeid_t` struct:
+
+```C
+        oe_sha256_context_t* hctx = &context->hash_context;
+        sgx_sigstruct_t* sigstruct = (sgx_sigstruct_t*)properties->sigstruct;
+        memcpy(eeid->sigstruct, (uint8_t*)sigstruct, sizeof(sgx_sigstruct_t));
+        eeid->vaddr = *vaddr;
+        eeid->entry_point = oeimage.entry_rva;
+
+        uint64_t ee_sz = sizeof(oe_eeid_t) + eeid->data_size;
+        uint64_t epg_sz = eeid_pages_size(eeid);
+        uint64_t num_pages = epg_sz / OE_PAGE_SIZE;
+        assert(*vaddr == enclave_end - epg_sz);
+
+        oe_page_t* pages = (oe_page_t*)eeid;
+        if (ee_sz < epg_sz)
+        {
+            oe_page_t* tmp = (oe_page_t*)calloc(1, epg_sz);
+            memcpy(tmp, eeid, ee_sz);
+            pages = tmp;
+        }
+
+        OE_CHECK(_add_extra_data_pages(
+            context, enclave->addr, pages, num_pages, vaddr));
+
+        if (ee_sz < epg_sz)
+            free(pages);
+
+        OE_SHA256 ext_mrenclave;
+        oe_sha256_final(hctx, &ext_mrenclave);
+```
+
+Finally, since our changes invalidate the signature of the base image, we need to dynamically re-sign the image. For this, we use the `OE_DEBUG_SIGN_KEY`, which is well-known by all OE users and thus does not require additional public-key infrastructure support. Note that applications should never rely on the apparent signing entity (e.g. `MRSIGNER`) of EEID-enabled evidence. Instead, what we provide to the application when we appraise EEID evidence, is the evidence of the base image.
+
+```C
+        OE_CHECK(oe_sgx_sign_enclave(
+            &ext_mrenclave,
+            properties->config.attributes,
+            properties->config.product_id,
+            properties->config.security_version,
+            OE_DEBUG_SIGN_KEY,
+            OE_DEBUG_SIGN_KEY_SIZE,
+            sigstruct));
+```
+
+We also add a serialization function `oe_serialize_eeid` for the `oe_eeid_t`, so that EEID can be sent to verifiers as an endorsement for verification of the base image evidence together with the EEID configuration of the instance.
+
+Verification of EEID reports
+----------------------------
+
+Verification of reports with EEID requires a fully populated `oe_eeid_t` and it performs the following steps (not necessarily in this order):
+
+- Create a SHA256 context and restore the internal state to `hash_state`,
+- re-measure the addition of heap, stack, and TCS pages with the settings from `size_settings` (starting at address `vaddr` and with `entry_point` in the TCS control pages),
+- measure the additional page containing the `oe_eeid_t` (filled with the entire `oe_eeid_t`, including `data`),
+- check that the final hash matches the hash reported in the extended report (e.g. MRENCLAVE),
+- check the identity of the signer of the extended image (e.g. public key corresponding to `OE_DEBUG_SIGN_KEY`),
```

Comment:
It's not clear to me how could we prevent a malicious program from hacking the EEID structure and re-sign it, since `OE_DEBUG_SIGN_KEY` is known to every one.

Reply:
It can attempt to do that, but remote attestation would fail. Since the EEID pages are measured, any change to them would make the enclave hash change. The debug key signature is not much more than an integrity check; it avoids distributing new keys. Note that the base image hash and signature (the whole sigstruct infact) are kept and verified, so nothing changes there.

Reply:
Please include the fact that the final mrenclave (and the whole sigstruct) is integrity protected by sgx hardware.

Reply:
Done!